File: Algorithms/CombSort.java
/*

Comb sort is a relatively simple sorting algorithm originally designed by Włodzimierz Dobosiewicz and Artur Borowy in 1980, later rediscovered by Stephen Lacey and Richard Box in 1991. Comb sort improves on bubble sort in the same way that Shellsort improves on insertion sort. Wikipedia
Inventor: Wlodzimierz Dobosiewicz, Stephen Lacey, Richard Box
Worst complexity: n^2
Average complexity: n*log(n)
Best complexity: n
Space complexity: 1

*/
// Java program for implementation of Comb Sort
// Available at: https://www.geeksforgeeks.org/comb-sort/
class CombSort
{
	// To find gap between elements
	int getNextGap(int gap)
	{
		// Shrink gap by Shrink factor
		gap = (gap*10)/13;
		if (gap < 1)
			return 1;
		return gap;
	}

	// Function to sort arr[] using Comb Sort
	void sort(int arr[])
	{
		int n = arr.length;

		// initialize gap
		int gap = n;

		// Initialize swapped as true to make sure that
		// loop runs
		boolean swapped = true;

		// Keep running while gap is more than 1 and last
		// iteration caused a swap
		while (gap != 1 || swapped == true)
		{
			// Find next gap
			gap = getNextGap(gap);

			// Initialize swapped as false so that we can
			// check if swap happened or not
			swapped = false;

			// Compare all elements with current gap
			for (int i=0; i<n-gap; i++)
			{
				if (arr[i] > arr[i+gap])
				{
					// Swap arr[i] and arr[i+gap]
					int temp = arr[i];
					arr[i] = arr[i+gap];
					arr[i+gap] = temp;

					// Set swapped
					swapped = true;
				}
			}
		}
	}

	// Driver method
	public static void main(String args[])
	{
		CombSort ob = new CombSort();
		int arr[] = {8, 4, 1, 56, 3, -44, 23, -6, 28, 0};
		ob.sort(arr);

		System.out.println("sorted array");
		for (int i=0; i<arr.length; ++i)
			System.out.print(arr[i] + " ");

	}
}
/* This code is contributed by Rajat Mishra */

File: Algorithms/comb_sort.py
"""
Comb Sort Algorithm

Comb Sort is an improvement over Bubble Sort. It uses a gap larger than 1,
which starts large and shrinks by a factor of 1.3 in each iteration.
The gap continues until it becomes 1, at which point it becomes a regular Bubble Sort.

Time Complexity:
    - Best Case: O(n log n)
    - Average Case: O(n² / 2^p) where p is number of increments
    - Worst Case: O(n²)

Space Complexity: O(1) - in-place sorting

Stable: No (may change relative order of equal elements)
"""


def comb_sort(arr):
    """
    Sorts an array using the Comb Sort algorithm.
    
    Args:
        arr: List of comparable elements to be sorted (modified in-place)
    
    Returns:
        None (sorts in-place)
    """
    n = len(arr)
    
    # Initialize gap
    gap = n
    
    # Initialize swapped as true to ensure loop runs
    swapped = True
    
    # Keep running while gap is more than 1 or last iteration caused a swap
    while gap != 1 or swapped:
        # Find next gap (shrink factor is 1.3)
        gap = int(gap / 1.3)
        if gap < 1:
            gap = 1
        
        # Initialize swapped as false to check if swap happened
        swapped = False
        
        # Compare all elements with current gap
        for i in range(0, n - gap):
            if arr[i] > arr[i + gap]:
                arr[i], arr[i + gap] = arr[i + gap], arr[i]
                swapped = True


# Example usage
if __name__ == "__main__":
    numbers = [64, 34, 25, 12, 22, 11, 90]
    print(f"Original array: {numbers}")
    
    comb_sort(numbers)
    print(f"Sorted array: {numbers}")
